src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -459,31 +400,80 @@ bool Connection::websocket_binary_message_received(util::Span<const char> data)
 }
 
 
-bool Connection::websocket_close_message_received(std::error_code error_code, StringData message)
+void Connection::websocket_error_handler()
 {
-    if (error_code.category() == websocket::websocket_close_status_category() && error_code.value() != 1005 &&
-        error_code.value() != 1000) {
-        m_reconnect_info.m_reason = ConnectionTerminationReason::websocket_protocol_violation;
+    m_error_received = true;
+}
 
-        constexpr bool try_again = true;
-        SessionErrorInfo error_info{error_code, message, try_again};
 
-        // If the server sends a websocket close message with code 1009, then it's because we've sent an
-        // UPLOAD message that is too large for the server to process. Simply disconnecting/reconnecting will not
-        // be sufficient because when we re-connect we'll just try to send the same bad upload message.
-        //
-        // Since the handling of this error happens at a layer below the standard `ERROR` message handling
-        // we need to synthesize an `ERROR` message-like error info here to client reset when this error
-        // is received.
-        if (error_code.value() == 1009) {
-            error_info.error_code = make_error_code(ProtocolError::limits_exceeded);
-            error_info.server_requests_action = ProtocolErrorInfo::Action::ClientReset;
-            error_info.message = util::format(
-                "Sync websocket closed because the server received a message that was too large: %1", message);
-        }
+bool Connection::websocket_closed_handler(bool was_clean, Status status)
+{
+    logger.info("Closing the websocket with status='%1', was_clean='%2'", status, was_clean);
+    auto&& status_code = status.code();
+    std::error_code error_code{static_cast<int>(status_code), websocket_close_status_category()};
 
+    // TODO: Use a switch statement once websocket errors have their own category in exception unification.
+    if (status_code == ErrorCodes::ResolveFailed || status_code == ErrorCodes::ConnectionFailed) {
+        m_reconnect_info.m_reason = ConnectionTerminationReason::connect_operation_failed;
+        constexpr bool try_again = true;
+        involuntary_disconnect(SessionErrorInfo{error_code, try_again}); // Throws
+    }
+    else if (status_code == ErrorCodes::ReadError || status_code == ErrorCodes::WriteError) {
+        read_or_write_error(error_code);
+    }
+    else if (status_code == ErrorCodes::WebSocket_GoingAway || status_code == ErrorCodes::WebSocket_ProtocolError ||
+             status_code == ErrorCodes::WebSocket_UnsupportedData || status_code == ErrorCodes::WebSocket_Reserved ||
+             status_code == ErrorCodes::WebSocket_AbnormalClosure ||
```

Comment:
Move the handling for `ErrorCodes::WebSocket_AbnormalClosure` to the last `else if` check, since for custom websocket implementations we won't be able to tell the difference between abnormal closure due to connection failed or a 401 error - at least not until the server has been updated to return the 401 in a websocket close code value.